readme.md
# README

Tool for calculating seasonal precipitation 

This project is licensed under the MIT licence. This project includes a copy of 
multigrids, and code based on  `atm.tools.initiation_areas.py` from the 
[atm project](https://github.com/ua-snap/arctic_thermokarst_model) which is 
licensed under the MIT licence. 


## Utility help
```
Utility for creating seasonal precipitation products by summing monthly
raster files. 

Flags
-----
    --monthly-data-directory: path
        input monthly raster data
    --out-directory: path
        Directory to save results at
    --season: string
        Season to sum precipitation for. Seasons are defined below 
        with inclusive month ranges
        'winter' for Oct(year N) - Mar(year N+1)
        'early winter' for Oct - Nov
        summer for Apr - Sept
    --sort-method: string, optional
        'default' to use Python's `sorted` method or,
        'snap' use a function to sort SNAP raster files which are
        named '*_MM_YYYY.*'

    
Example
-------
    python spc/utility.py --monthly-data-directory=./ --out-directory=./ 
        --season=winter --sort-method=snap
```


changelog.md
# Changelog

## 3.0.0 [2022-02-08]
### added
- multiprocessing support
- limited restart capabilities
- added support for summer seasons

## 2.0.0 [2021-11-03]
### added
- tools and cli options to calculate seasonal precipitation based on roots 
(representing change from thawing to freezing or vice-versa) output from the 
degree-day calculator

## 1.0.1 [2020-10-09]
First release  


spc/CLILib.py
"""
CLI Lib
-------

Simple python cli wrapper

Rawser Spicer

taken from utilitools v0.1.0

version: 1.0.0
Split from imiq-database python utilities: 2017-08-18

"""
import sys

class CLILibTypeError(Exception):
    """CLILib type error, raised if type is wrong for a provided flag
    """
    pass
    
class CLILibMandatoryError(Exception):
    """CLILib mandatory flag error, raised if all of the mandatory flags 
    are not present
    """
    pass
    
class CLILibHelpRequestedError(Exception):
    """CLILib help requested error, raised if -h or --help is provied in the 
    lsit of flags
    """
    pass

class CLI (object):
    """ This class will act as a simple CLI utility """
    
    def __init__ (self, mandatory, optional = [], types = {}):
        """This class will act as a simple CLI utility
        
        Parameters
        ----------
        mandatory: list
            list of mandatory flags for utility
        optional: list, optional
            list of optional flags for utility
        types: dict, optional
            dictioanry of each flag with the type it should be 
            
        .. note:: -h or --help are reserved to rasis help requested error
    
        attributes
        ----------
        flags: list
            lais of all flags for utility
        args: dict
            dictionary of each flag with its values
            
        raises
        ------
        CLILibHelpRequestedError 
            if -h or --help are provided in list of flags
        
        """
        self.flags = mandatory + optional
        # print(self.flags)
        self.args = {}
        
        if '-h' in sys.argv[1:] or '--help' in sys.argv[1:]:
            raise CLILibHelpRequestedError
        
        for item in sys.argv[1:]:
            try:
                flag, value = item.split('=')
                self.args[flag] = value
            except ValueError:
                flag = item.split('=')[0]
                self.args[flag] = True
        if not set(mandatory) <= set(self.args.keys()) or \
            not set(self.args.keys()) <= set(self.flags):
            raise CLILibMandatoryError("Invalid mandatory flags")
            
        
        if len(types) != 0:
            for flag in types:
                try:
                    self.args[flag] = types[flag]( self.args[flag] )
                except ValueError:
                    msg = 'Type of argument ' + flag + \
                        ' must be ' + str(types[flag]) + '.'
                    raise CLILibTypeError (msg)
                except KeyError:
                    pass
            
    def __repr__ (self):
        """
        """
        return str(self.args)
        
    def __getitem__ (self, key):
        """get opperator for utility, gets a flags value
        
        parameters
        ----------
        key: str
            the flag to get the value for
        
        returns
        -------
        the utilitys flags value
        """
        try:
            return self.args[key]
        except KeyError:
            if key in self.flags:
                return None
            else:
                raise KeyError(key)
        
        
        
        

if __name__ == "__main__":
    """ example utility """
    try:
        test = CLI(['--a','--b'],['--c'],
            types = {'--a':int, '--b': float, '--c':str})
        print (type(test['--a']), test['--a'])
        print (type(test['--b']), test['--b'])
        print (type(test['--c']), test['--c'])
    except  CLILibMandatoryError:
        print ("Valid flags are: --a, --b, and --c(optional)")
    except  CLILibTypeError as E:
        print (E)
        
    
    


spc/seasonal.py
"""
Seasonal Precipitation Calculator
---------------------------------



"""
from multigrids.tools import load_and_create, get_raster_metadata, temporal_grid
import glob, os
import re
import numpy as np
from datetime import datetime, timedelta

from multiprocessing import Process, Lock, active_children, cpu_count

def load_monthly_data (
        directory, dataset_name = 'Monthly Precipitation', sort_func = sorted,
        no_data = -9999 
    ):

    if os.path.isfile(directory):
        monthly_data = temporal_grid.TemporalGrid(directory)
        key = list(monthly_data.config['grid_name_map'].keys())[0]
        start = [int(i) for i in key.split('-')][0]
    else:
        rasters =  sort_func(glob.glob(os.path.join(directory,'*.tif')))

        match = r'([1-3][0-9]{3})'
        start = os.path.split(rasters[0])[-1]
        start = re.search(match, start).group(0)   
        end = os.path.split(rasters[-1])[-1]
        end = re.search(match, end).group(0)   

        grid_names = [] 
        for yr in range(int(start), int(end)+1): 
            for mn in ['01','02','03','04','05','06','07','08','09','10','11','12']: 
                grid_names.append(str(yr) + '-' + str(mn)) 

        load_params = {
            "method": "tiff",
            "directory": directory,
            "sort_func": sort_func,
        }
        create_params = {
            "name": dataset_name,
            "grid_names": grid_names,
        }

        monthly_data = load_and_create(
            load_params, create_params
        )

        ex_raster = rasters[0]

        monthly_data.config['raster_metadata'] = get_raster_metadata(ex_raster)

        ## precip is never less than zero TODO is this the cause of the weird issues?
        monthly_data.grids[monthly_data.grids < 0] = np.nan

    monthly_data.config['start_year'] = int(start)

    return monthly_data

def sum_precip(monthly, months, years='all', name="Summed-Precip"):
    """

    Parameters
    ----------
    monthly: multigrids.temporal_gird.TemporalGrid
        monthly precip data

    months: list of ints
        list of months numbers as integers where January is 1, February is 2, 
        and so on. If a season spans two years as winter does, use 13 as 
        the next January and 14 as the next February etc. So for a winter 
        comprising october through march pass [10, 11, 12 ,13 ,14 ,15].
    years: str
        Range of years to calculate average over. 
        'all' or 'start-end' ie '1901-1950'.
    
    Returns
    ------- 
    # winter_precip
    """
    keys = monthly.config['grid_name_map'].keys()

    find = lambda x: sorted( [k for k in keys if k[-2:]=='{:0>2}'.format(x)])
    monthly_keys = {
        x: find(x) for x in range(1, 13)
        }     

    pad = 1 - max([m//12 for m in months])

    if years != 'all':
        try:
            start, end = years.split('-')
        except AttributeError:
            start, end = years[0], years[1]
        start, end = int(start), int(end)
    else:
        start = min([int(x.split('-')[0])for x in keys]) 
        end = max([int(x.split('-')[0])for x in keys]) 
    

    month_filter = lambda m, y, ks: [k for k in ks[m] if k[:4] == str(y)]
    months_filtered = {}
    for year in range(start, end+pad):
        for mon in months:
            if mon < 1:
                raise IndexError ("Months cannot be less then 1")
            adj_mon = mon
            adj_year = year
            while adj_mon > 12:
                adj_mon -= 12
                adj_year += 1

            try:
                months_filtered[mon] +=  month_filter(
                    adj_mon, adj_year, monthly_keys
                ) 
            except KeyError:
                months_filtered[mon] = month_filter(
                    adj_mon, adj_year, monthly_keys
                )
    
    precip_sum = None
    for mon in months_filtered:
        try:
            precip_sum += monthly.get_grids_at_keys(months_filtered[mon])
        except TypeError:
            precip_sum = monthly.get_grids_at_keys(months_filtered[mon])

    
    
    precip_sum = temporal_grid.TemporalGrid(
        precip_sum.shape[1], # rows
        precip_sum.shape[2], # cols
        precip_sum.shape[0], # years
        data_type = precip_sum.dtype, 
        dataset_name = name,
        description = "summed precip for " + str(months),
        mask = monthly.config['mask'],
        initial_data = precip_sum,
        start_timestep = monthly.config['start_year'],
        )

    return precip_sum

def get_start_next_month(date):
    """
    Parameters
    ----------
    date: datetime.datetime

    Returns
    -------
    datetime.datetime
        date of the start of the next month
    """
    if date.month+1 <= 12:
        return datetime(date.year,date.month+1,1)  
    return datetime(date.year+1,1,1) ## its next year

def sum_element_by_date_range(monthly, start, end, row, col):
    """sum a season based on start and end dates. Fractional protion of months
    is calculated.

    Parameters
    ----------
    monthly: TemporalGrid
        monthly precip data
    start: np.datetime64
        start date to create seasonal sum
    end: np.datetime64
        end date to create seasonal sum
    row: int
    col: int
        row and col to process
    
    Returns
    -------
    float
        sum of seasonal precip 
    """
    try:
        start = datetime.fromisoformat(str(start))
    except ValueError:
        return np.nan
    end = datetime.fromisoformat(str(end))

    keys = []
    month = start.month
    year = start.year
    while True:
        keys.append('%i-%02i' % (year,month))

        if year == end.year and month == end.month:
            break

        month += 1
        if month == 13:
            month = 1
            year += 1


    data = monthly.get_grids_at_keys(keys)[:, row,col]
    if len(data) == 1:
        ## do somthing else:
        total_days = (
            get_start_next_month(end) - datetime(end.year,end.month,1)
        ).days
        days = (end.day + 1) - start.day
        fraction = days/total_days
        return data[0] * fraction 

    total_days = (
        get_start_next_month(start) - datetime(start.year,start.month,1)
    ).days
    days = (get_start_next_month(start) - start).days
    fraction_start = days/total_days
    data[0] = data[0] * fraction_start    
    
    total_days = (
        get_start_next_month(end) - datetime(end.year,end.month,1)
    ).days
    days = end.day
    fraction_end = days/total_days
    data[-1] = data[-1] * fraction_end

    return data.sum()


def process_pixel_warper(monthly, bounds, row, col, start_year, grids, lock ):
    """multiprocessing wrapper to process a single year for a single pixel

    Parameters
    ----------
    monthly: TemporalGrid
        monthly data
    bounds: 
    row: int
    col: int
        row, and col to process
    start_year: int  
        year to start processing at
    grid: np.array like
        array to store the results to
    lock: multiprocessing.Lock
        Lock for writing data

    """
    year = start_year

    for bound_pair in bounds:
        start, end = bound_pair[0],bound_pair[1]
        if lock:
            lock.acquire()
        grids[year, row, col] = sum_element_by_date_range(
            monthly, start, end, row, col
        )
        if lock:
            lock.release()
        year += 1


def sum_grids_by_ranges(monthly, dates, n_days=None, num_process = 1, 
        start_row=0, start_col =0, end_row = None,
        summed_data = None
    ):
    """
    calculates seasonal precip based on date ranges

    Parameters
    ----------
    monthly: TemporalGrid
        monthly precip data
    dates: TemporalGrid
        dates grid with alternating freezing then thawing dates
    n_days: int, Optional
        override for season length, if this is used the end date for 
        the seasion is the thawing data my n_days from freezing date.
        Full seaasons are used if this is not provided
    num_processes: int default 
        number of process for multiprocessing
    start_row: int
    start_col: int
        start row and start columns to use
    end_row: int, Optional
        if none number of rows in monthly data is used
    summed_data: np.array or TemporalGrid
        if None data is created, otherwise data is stored in array or 
        TemporalGrid provided
    
    Returns
    -------
    np.array with shape (n_year, row, cols)
    """
    w_lock = Lock()
    if num_process is None:
       num_process = cpu_count()

    grid_shape = monthly.config['grid_shape']
    n_years = dates.config['num_grids']//2
    shape = (n_years,grid_shape[0],grid_shape[1])

    if summed_data is None: ## create resulting data space if none exists
        filename = 'calc_precip_temp.data'
        grids = np.memmap(filename, dtype='float32', mode='w+', shape=shape) 
        grids += np.nan
    else:
        try:
            grids = summed_data.grids
        except AttributeError:
            grids = summed_data
    
    real_shape = dates.config['real_shape']
    bounds_shape = (dates.config['num_timesteps']//2,2)
    disp_bounds = True
    
    if end_row is None:
        rows = range(start_row, grid_shape[0])
        end_row = grid_shape[0]
    else:
        rows = range(start_row, end_row)

    for row in rows:
        print (row, '/', grid_shape[0], ':', row/(end_row-start_row)*100 )
        for col in range(start_col, grid_shape[1]):
            if np.isnan(monthly[0][row,col]):
                continue
            
            bounds = dates.grids.reshape(real_shape)[:,row,col]\
                .reshape(bounds_shape)
            if n_days:
                bounds[:,1] = bounds[:,0] + np.timedelta64(n_days,'D')
            if disp_bounds:
                print(bounds)
                disp_bounds=False

            year_idx = 0

            if num_process > 1: 
                while len(active_children()) >= num_process:
                    continue
                # (monthly, bounds, row, col, start_year, grids, lock ) <- args
                Process(target=process_pixel_warper,
                        name = "Calc precip for all years in r:%i, c:%i" % (row, col),
                        args=(
                            monthly, bounds, row, col, year_idx, grids, w_lock
                        )
                    ).start()
            else: ## no multiprocessing
                process_pixel_warper(
                    monthly, bounds, row, col, year_idx, grids, None
                )

    if num_process > 1:
        print('Waiting on final processes to complete')
        while len(active_children()) > 0 :
            
            continue
        print('Done')

    return grids

def sum_seasonal_2(monthly, dates, n_days=None, season='winter', start_year=0, 
        name="Summed-Precip"
    ):
    """calculates seasonal precip based on freezing/thawing dates

    parameters
    ----------
    monthly: TemporalGrid
        monthly precip data
    dates: TemporalGrid
        dates grid with alternating freezing then thawing dates
    n_days: int or string, optional
        override for season length, if this is used the end date for 
        the seasion is the thawing data my n_days from freezing date

        if a string 'summer' or 'winter' should be provided to calculate the
        season from the roots for the respective season 
    start_year: int, default 0
        start year for labeling 
    name: str
        name used for labeling

    returns
    -------
    TemporalGrid
        the seasonal precip

    """
    precip_sum = sum_grids_by_ranges(monthly, dates, n_days, num_process=8) 

    precip_sum = temporal_grid.TemporalGrid(
        precip_sum.shape[1], # rows
        precip_sum.shape[2], # cols
        precip_sum.shape[0], # years
        data_type = precip_sum.dtype, 
        dataset_name = name,
        description = "seasonal precip summed by date ranges",
        mask = monthly.config['mask'],
        initial_data = precip_sum,
        start_timestep = start_year,
        )

    return precip_sum


def root_mg_to_date_mg(roots, start_date, skip_at_start=0, skip_at_end=None):
    """convert TemporalGrid of roots to TemporalGrid of dates

    Parameters
    ----------
    roots: TemporalGrid
        the roots data, where -numbers represent freezing roots and + numbers
        thawing roots. The absolunte value of each root is the number of 
        days from the start_date
    start_date: datetime.datetime
        day '0' for roots
    skip_at_start: int default 0
        number of roots to skip at start
    skip_at_end: int, default None
        number of roots to skip at end

    Returns
    -------
    TemporalGrid
        type is np.datetime64 
    """

    skip = skip_at_start + skip_at_end if skip_at_end else 0

    dates_mg = temporal_grid.TemporalGrid(
        roots.config['grid_shape'][0], # rows
        roots.config['grid_shape'][1], # cols
        roots.config['num_timesteps'] - skip, # years
        data_type = 'datetime64[D]', 
        start_timestep = 0,
        )

    dates_mg.grids[:] =  np.datetime64(start_date - timedelta(days=1)) 
    dates_mg.grids[:] += abs(roots.grids).astype(np.timedelta64)[
        skip_at_start:( -1 * skip_at_end) if skip_at_end else None
    ] 
    return dates_mg




spc/utility.py
"""
Seasonal Precipitation Calculator Utility 
-----------------------------------------

Utility for creating seasonal precipitation products by summing monthly
raster files. 

"""
from multigrids.temporal_grid import TemporalGrid
from multigrids import tools
import CLILib 


import os, sys
import numpy as np

import seasonal 

from sort import sort_snap_files

from datetime import datetime

SEASONS = {
    "winter": [10, 11, 12, 13, 14, 15],
    "early-winter": [10, 11],
    "summer": [4, 5, 6, 7, 8, 9],
    "late-summer": [8, 9],
}


def utility ():
    """
    Utility for creating seasonal precipitation products by summing monthly
    raster files. 

    Flags
    -----
        --monthly-data: path
            input monthly raster data
        --out-directory: path
            Directory to save results at
        --method: string
            monthly, or roots
            if monthly, --season must be used also
            if roots,  --roots-directory or  --roots-file must be used also
                --season-length may also be used with roots method
        --season: string
            Season to sum precipitation for. Seasons are defined below 
            with inclusive month ranges
            'winter' for Oct(year N) - Mar(year N+1)
            'early winter' for Oct - Nov
            summer for Apr - Sept

        --roots-directory:
            directory containing roots data in tiff format in files that
            can be sorted correctly using pythons sorted methdod
        --roots-file:
            roots file multigird
        --season-length: int
            length in days of season to use with roots.  
        --out-format:
            tiff, or multogrid
        --sort-method: string, optional
            sort method of input monthly precipitation data
            'default' to use Python's `sorted` method or,
            'snap' use a function to sort SNAP raster files which are
            named '*_MM_YYYY.*'
        
        
    Example
    -------
        python spc/utility.py --monthly-data=./ --out-directory=./ 
            --season=winter --sort-method=snap
    """
    try:
        arguments = CLILib.CLI(
            ['--monthly-data', '--out-directory'],
            [
            '--sort-method', '--season', '--method',
            '--roots-directory', '--roots-file',
            '--season-length', '--out-format', 
            ]   
        )
    except (CLILib.CLILibHelpRequestedError, CLILib.CLILibMandatoryError) as E:
        print (E)
        print(utility.__doc__)
        return

    if arguments['--sort-method']is None:
        sort_method = "Using default python sort function"
        sort_fn = sorted
    elif  arguments['--sort-method'].lower() == 'snap':
        sort_method = "Using SNAP sort function"
        sort_fn = sort_snap_files
    elif not arguments['--sort-method']is None and\
        arguments['--sort-method'].lower() != "default":
        print("invalid --sort-method option")
        print("run utility.py --help to see valid options")
        print("exiting")
        return

    monthly = seasonal.load_monthly_data(
        arguments['--monthly-data'], sort_func=sort_fn
    )

    if arguments['--method'] is None or arguments['--method'] == 'monthly':
        season = arguments['--season']
        name = season + '-precipitation'
        summed = seasonal.sum_precip(monthly, SEASONS[season], name=name)

        summed.config['description'] = name + " for months " +\
            str(SEASONS[season]) +\
            "(month '13' is january for the next year, '14' is feb, etc)"
        summed.config['raster_metadata'] = monthly.config['raster_metadata']
    elif arguments['--method'] == 'roots':

        if arguments['--roots-file']:
            print(' roots file')
            roots = TemporalGrid(arguments['--roots-file'])
        elif arguments['--roots-directory']:
            load_params = {
            "method": "tiff",
            "directory": arguments['--roots-directory'],
            "sort_func": sorted
            }
            create_params = {
                "name": 'roots',
                "start_timestep": 0,
            }

            roots = tools.load_and_create(
                load_params, create_params
            )
        else:
            print ("'--roots-file' or '--roots-directory' must be"
                   " suppied for 'roots' method"
            )

        key = list(monthly.config['grid_name_map'].keys())[0]
        # print(key)
        year, month = [int(i) for i in key.split('-')]
        start_date = datetime(year, month, 1)
        # print(start_date)

        try:
            season_name, season_length = arguments['--season-length'].split('-')
        except ValueError:
            season_name = arguments['--season-length']
            season_length = None

        if season_name == 'summer':
            dates = seasonal.root_mg_to_date_mg(roots, start_date) # don't need to skip ends
        elif  season_name == 'winter':
            dates = seasonal.root_mg_to_date_mg(roots, start_date, 1, 1)
        # print(dates)

        # try:
        monthly.config['start_timestep'] = 0
        print(monthly.config['description'])
        summed = seasonal.sum_seasonal_2(
            monthly, dates, season_length, season_name
        ) 
        # except IndexError as e:
        #     print (e)
        #     sys.exit()

        summed.config['start_timestep'] = year
        summed.config['raster_metadata'] = monthly.config['raster_metadata']

        # import matplotlib.pyplot as plt
        # plt.imshow(summed[1950])
        # plt.colorbar()
        # plt.show()

    else:
        print ("'--method' invalid -> %s" % str(arguments['--method']))




    try:
        os.makedirs(arguments['--out-directory'])
    except:
        pass
    if arguments['--out-format'] is None or \
            arguments['--out-format'] == 'tiff' :
        summed.save_all_as_geotiff(arguments['--out-directory'])
    elif arguments['--out-format'] == 'multigrid':
        summed.config['command-used-to-create'] = ' '.join(sys.argv)
        summed.save(
            os.path.join(
                arguments['--out-directory'], 
                'seasonal-precip.yml'
            )
        )
    


utility()
